src/peft/tuners/poly/model.py

```diff
@@ -0,0 +1,151 @@
+from dataclasses import asdict
+from enum import Enum
+from typing import Any
+
+import torch
+from torch import nn
+
+from peft.tuners.tuners_utils import BaseTuner, BaseTunerLayer, check_target_module_exists
+from peft.utils import (
+    TRANSFORMERS_MODELS_TO_LORA_TARGET_MODULES_MAPPING,
+    ModulesToSaveWrapper,
+)
+
+from .config import PolyConfig
+from .layer import Linear, PolyLayer
+
+
+class PolyModel(BaseTuner):
+    prefix: str = "poly_"
+
+    def __init__(self, model, config, adapter_name) -> None:
+        super().__init__(model, config, adapter_name)
+        self.register_forward_pre_hook(PolyModel._add_task_id_pre_hook, with_kwargs=True)
```

Comment:
you are right, having the handle alive here could be annoying. it would be better to clean it up after using, maybe with a (post-)forward hook or similar.